### PR TITLE
[Merged by Bors] - feat: `interval_cases` tactic

### LINKED ISSUE #1110: `interval_cases` tactic
Hopefully not too hard?  In mathlib3 this is a 300-line wrapper for `fin_cases` ....

### PR DESCRIPTION
**Pull request by digama0**
Implements the `interval_cases` tactic from mathlib. This is a from scratch implementation, which should be a lot faster and give better goals than the original. I also tried to make a reasonable MetaM version of the tactic.

The original tactic was very generic and used typeclass inference to find most of the structure on the type. The new one is still somewhat type-generic but relies on meta-instances which implement the required functionality for supported types. Currently only `Nat` and `Int` are supported; `PNat` is more difficult because `norm_num` doesn't work on it.

### CODE REVIEW

File: Mathlib/Tactic/IntervalCases.lean [hunk 1]
```diff
@@ -0,0 +1,282 @@
+/-
+Copyright (c) 2022 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Heather Macbeth
```
Comment (hrmacbeth) [inline Mathlib/Tactic/IntervalCases.lean:4]: I have never been involved with this tactic, maybe this is copy-pasted from another header?

**Comment by semorrison** [comment]
bors merge

**Comment by hrmacbeth** [comment]
The new `Rat` functionality of `norm_num` broke the build, and I have fixed it (at least, it builds) by blindly replacing `toRawEq`s with `toRawIntEq.get!`s, following the model of the `ring` fixes in #1441.  @digama0 Feel free to revert if I did it wrong.

**Comment by digama0** [comment]
bors merge

**Pull request closed without merging**